EDoc: EEP-48 doc chunk generation

erszcz opened the pull request:
This PR enables EDoc to output EEP-48 style doc chunks. The work leading to this PR was so far discussed and documented at https://github.com/erlef/documentation-wg/issues/4, whereas developed at https://github.com/erszcz/edoc.

I tried to address all points raised in the discussions so far:
1. Chunks are written to disk as standalone files, not embedded into `.beam` files.
1. It's possible to generate chunks from the command line by running a script.
1. The chunks are as close to `erl_docgen` chunks as possible (with the caveat of type links described below).
1. All `@spec` and `@type` tags are rewritten to `-spec` and `-type` attributes, respectively.

We discussed the type link caveat with @garazdawi - `erl_docgen` chunk links to types do not use type arities. This means links to `t/0` and `t/1` have the same form. EDoc chunks do not follow this principle, since it seemed more robust to allow linking to arbitrary types. Until this is fixed for `erl_docgen` chunks, measures have to be taken in 3rd party tools (e.g. ExDoc) to watch for this difference.

Some features stemming from EDoc design and implementation, but not necessarily intuitive, are not addressed in this PR:
1. Function doc comments can precede or follow `-spec` attributes, but type doc comments must follow `-type` attributes.
1. `@since` and `@deprecated` tags are supported on functions, but are not supported on types or callbacks.

This PR also adds `test/eep48_SUITE.erl` which tests the majority of the newly added functionality, as well as some of the above points which are currently not supported (these tests are skipped by default).

josevalim commented:
This is amazing work @erszcz! :heart:

One of the things to discuss, probably for after this PR is merged, is to deprecate the tags `@spec` and `@type` in EDoc. Possibly stop parsing them altogether and emit a warning, since they won't be used by chunks and, afaik, the offer not benefit (only downsides?) over the existing `-type` and `-spec`.

KennethL commented:
I was just about to ask if we could get this in place for the OTP 24
release (and release candidates beginning in February) so this is fantastic
good news.
Thanks @erszcz.

/Regards Kenneth

On Wed, Oct 14, 2020 at 5:09 PM Radek Szymczyszyn <notifications@github.com>
wrote:

> This PR enables EDoc to output EEP-48 style doc chunks. The work leading
> to this PR was so far discussed and documented at erlef/documentation-wg#4
> <https://github.com/erlef/documentation-wg/issues/4>, whereas developed
> at https://github.com/erszcz/edoc.
>
> I tried to address all points raised in the discussions so far:
>
>    1. Chunks are written to disk as standalone files, not embedded into
>    .beam files.
>    2. It's possible to generate chunks from the command line by running a
>    script.
>    3. The chunks are as close to erl_docgen chunks as possible (with the
>    caveat of type links described below).
>    4. All @spec and @type tags are rewritten to -spec and -type
>    attributes, respectively.
>
> We discussed the type link caveat with @garazdawi
> <https://github.com/garazdawi> - erl_docgen chunk links to types do not
> use type arities. This means links to t/0 and t/1 have the same form.
> EDoc chunks do not follow this principle, since it seemed more robust to
> allow linking to arbitrary types. Until this is fixed for erl_docgen
> chunks, measures have to be taken in 3rd party tools (e.g. ExDoc) to watch
> for this difference.
>
> Some features stemming from EDoc design and implementation, but not
> necessarily consistent with expectations towards a modern documentation
> system are not addressed in this PR:
>
>    1. Function doc comments can be precede or follow -spec attributes,
>    but type doc comments must follow -type attributes.
>    2. Callbacks cannot have associated doc comments at all.
>    3. @since and @deprecated tags are supported on functions, but are not
>    supported on types or callbacks.
>
> This PR also adds test/eep48_SUITE.erl which tests the majority of the
> newly added functionality, as well as some of the above points which are
> currently not supported (these tests are skipped by default).
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/erlang/otp/pull/2803
> Commit Summary
>
>    - Import EEP-48 chunk support
>    - Merge branch 'import-edoc-chunks' into edoc-chunk-support
>    - Fix build issues with warnings-as-errors
>    - Update SOURCES in src/Makefile
>    - Fix more warnings-as-errors build issues
>    - Enable eep48_SUITE
>    - Output callback chunk entries based on #tag{} records
>    - Describe EEP-48 chunk support and how to generate chunks with EDoc
>    - Drop the EDoc chunk Rebar3 plugin from this repository
>    - Fix some types
>    - Export shell_docs:{chunk_elements,chunk_element_attr}/0
>    - Add edoc_doclet to erl_docgen's list of modules known not to have
>    any functions
>    - Fix Dialyzer errors
>    - Remove some TODOs
>
> File Changes
>
>    - *A* lib/edoc/bin/edoc.escript
>    <https://github.com/erlang/otp/pull/2803/files#diff-ee778b4af2631a1e1aa6984ef56ba35e6ffab41beeb7025d7d92fba048623c7d>
>    (5)
>    - *M* lib/edoc/doc/overview.edoc
>    <https://github.com/erlang/otp/pull/2803/files#diff-7bc60eb4edf1ff4080c56a92eb534fa9f00b3e21ba68694ba6b1dc1882d53101>
>    (39)
>    - *M* lib/edoc/include/edoc_doclet.hrl
>    <https://github.com/erlang/otp/pull/2803/files#diff-a8cd5451e7912fb1caf64214b20bc69a778ef41502cfa1ed014073d78f9127e9>
>    (37)
>    - *M* lib/edoc/src/Makefile
>    <https://github.com/erlang/otp/pull/2803/files#diff-21c03813315afd3004e0fc4c9aabe1bac8f6330a22a80e91c140e0fa6b030241>
>    (8)
>    - *M* lib/edoc/src/edoc.app.src
>    <https://github.com/erlang/otp/pull/2803/files#diff-0a87902b12eb6aa05129af33e7b184453f98012d7c9eee9f4b5f6c7aca6b7c04>
>    (6)
>    - *M* lib/edoc/src/edoc.erl
>    <https://github.com/erlang/otp/pull/2803/files#diff-0b079f122092e17e86f314df15ba27691ff93d9316ea7e724b8bcf260e5a4098>
>    (171)
>    - *M* lib/edoc/src/edoc.hrl
>    <https://github.com/erlang/otp/pull/2803/files#diff-977c7f72433598c4ac43ab02af3242866076c32acbd2c8f39e95398cf694a7a0>
>    (45)
>    - *A* lib/edoc/src/edoc_cli.erl
>    <https://github.com/erlang/otp/pull/2803/files#diff-b3ef7e087b1a2ad92ac1e72fd39af8efb7574f69f3e3650595e071aa0e70b189>
>    (95)
>    - *M* lib/edoc/src/edoc_data.erl
>    <https://github.com/erlang/otp/pull/2803/files#diff-4d0162619716552eb4c6c2f16d3e0f03eb143e8a75517b7a4d27e4a7594eb513>
>    (35)
>    - *M* lib/edoc/src/edoc_doclet.erl
>    <https://github.com/erlang/otp/pull/2803/files#diff-fba37bba9fd6411f788c8864b0a042d6f4e844db53b4b4d7b0042e8a2b09fd67>
>    (441)
>    - *A* lib/edoc/src/edoc_doclet_chunks.erl
>    <https://github.com/erlang/otp/pull/2803/files#diff-e4410b8280b27cf26217699e9fa9c536829181db786b570802bcd2a40dbe11b9>
>    (117)
>    - *A* lib/edoc/src/edoc_doclet_default.erl
>    <https://github.com/erlang/otp/pull/2803/files#diff-590d21cc001fcba0ce54ffc5381911546b448c24607ba0356052946a08b4f636>
>    (462)
>    - *M* lib/edoc/src/edoc_extract.erl
>    <https://github.com/erlang/otp/pull/2803/files#diff-188f547a2e63e6d2b93b4b634d27284bd8f532995710f4c7baaa5a66c7a08a78>
>    (248)
>    - *M* lib/edoc/src/edoc_layout.erl
>    <https://github.com/erlang/otp/pull/2803/files#diff-7ceb762dc31de77e955cdc811dcb2c760b0c5c49d5d121aacb034fd9f0e04ba3>
>    (12)
>    - *A* lib/edoc/src/edoc_layout_chunks.erl
>    <https://github.com/erlang/otp/pull/2803/files#diff-cc2ad3d2874b85f201ad812ada23b33f3fa4cdc92a790b4dc78c76d4182ac708>
>    (393)
>    - *M* lib/edoc/src/edoc_lib.erl
>    <https://github.com/erlang/otp/pull/2803/files#diff-9d4fb29b418b494e1cbf7de1c7340fb4f5b63cc409b23cf6e0898e99cdbf2206>
>    (63)
>    - *M* lib/edoc/src/edoc_macros.erl
>    <https://github.com/erlang/otp/pull/2803/files#diff-7874d8c1279db0d58e35f8dabda5931598cf2d140f70d10b4aa57d27f8ad8a7f>
>    (5)
>    - *M* lib/edoc/src/edoc_parser.yrl
>    <https://github.com/erlang/otp/pull/2803/files#diff-08fac87126bc3f83aaee7fb3631289ff69898c01f5c25a5374f36e5915b87faf>
>    (12)
>    - *M* lib/edoc/src/edoc_refs.erl
>    <https://github.com/erlang/otp/pull/2803/files#diff-656cace0664d499b8847f3949b440d03b59cb6b38c7bcfbaf15d53f2acc361b9>
>    (47)
>    - *M* lib/edoc/src/edoc_run.erl
>    <https://github.com/erlang/otp/pull/2803/files#diff-41a5d13b5a98a97c5faf0f614d5ca267ec751c8f3542e021562397aec5fa342b>
>    (18)
>    - *M* lib/edoc/src/edoc_specs.erl
>    <https://github.com/erlang/otp/pull/2803/files#diff-2d7bbab975312d3c634bd173db9f6567021edb5057af5407ab3801a983a99577>
>    (59)
>    - *M* lib/edoc/src/edoc_tags.erl
>    <https://github.com/erlang/otp/pull/2803/files#diff-e91e9acf3ba1063cb68d62e18ea200a80fcad988264be31af5cbdbf9e249e6ed>
>    (20)
>    - *M* lib/edoc/src/edoc_types.erl
>    <https://github.com/erlang/otp/pull/2803/files#diff-e96bd7c16c6eb37adce5c88fe687fde14e793abd170e93e8691c4bfb3a1c4e8e>
>    (94)
>    - *M* lib/edoc/src/edoc_types.hrl
>    <https://github.com/erlang/otp/pull/2803/files#diff-3cd9a87a9a2b1f736a5ac3fe9b417052666cc896900508edb6601e52e0b88686>
>    (75)
>    - *M* lib/edoc/test/Makefile
>    <https://github.com/erlang/otp/pull/2803/files#diff-107c51805818c42250d762d26b196e0d782d6526d9db1d4ed26eaf7e68c9a368>
>    (3)
>    - *A* lib/edoc/test/eep48_SUITE.erl
>    <https://github.com/erlang/otp/pull/2803/files#diff-4cca9d805c2e58708406fd194dd1640ce9f66f27404c10ec99acf0c7cc757291>
>    (290)
>    - *A* lib/edoc/test/eep48_SUITE_data/eep48_links.erl
>    <https://github.com/erlang/otp/pull/2803/files#diff-f84ff0e291adc4bc1ec4ced67efdde94043057e5fc632d383dbb07a7e237e033>
>    (102)
>    - *A* lib/edoc/test/eep48_SUITE_data/eep48_meta.erl
>    <https://github.com/erlang/otp/pull/2803/files#diff-65229d9ea93b2e0bb1b18c42b7507849a2b3bf76b0a1edb9b79bf737c5dc867a>
>    (35)
>    - *M* lib/erl_docgen/src/docgen_xml_to_chunk.erl
>    <https://github.com/erlang/otp/pull/2803/files#diff-7c882197ab5fc0a16a579261802a2d72e18a52560e68ccf2e4ccf71f4415a82a>
>    (3)
>    - *M* lib/stdlib/src/shell_docs.erl
>    <https://github.com/erlang/otp/pull/2803/files#diff-f0464950a78b61375e01dc0c9f31222f83dd71d85c77783bd39d9e57379a3f0e>
>    (2)
>
> Patch Links:
>
>    - https://github.com/erlang/otp/pull/2803.patch
>    - https://github.com/erlang/otp/pull/2803.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/erlang/otp/pull/2803>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABFWSFVSGFVHFSGAJD4LRLSKW5K5ANCNFSM4SQXUSYQ>
> .
>

garazdawi reviewed (commented):
I've done a cursory review and I think all in it it looks very good. Most of the comments are about backward-compatibility and documentation.

It would be nice if the changes of `@spec` to `-spec` would be a separate commit from everything else so that it is easier to see that it does not change any functionality.

Does @richcarl have any comments about this PR?

I also looked through some rendered functions and found some small irregularities that I'm not sure if you have seen or not:

### erl_parse:tree/2

In the edoc html generate this function looks like this:

    tree(Type::atom(),Data::term()) -> tree()

however, in the shell docs it looks like this:

    -spec tree(atom(),term()) -> tree().

This because edoc for html reads the variable names and inserts them
into the spec. Information is lost when this is not done as the reader
does not know which argument is `Type` and which is `Data` in the text.

    -spec tree(atom(), term()) -> tree().
    
    For special purposes only. Creates an abstract syntax tree node
    with type tag *Type* and associated data *Data*.

### See also

    See also data/1.
    
    See also is_tree/1.
    
    See also type/1.

These should probably be grouped into one:

    See also data/1, is_tree/1, type/1

as that is done both in the edoc html and shell_docs.

### Problem in shell_docs

Seems like shell_docs does not render `<pre>` correctly, or rather
it does not insert a newline before itself so that things like merl
look like this:

```
  Quick start

  To enable the full power of Merl, your module needs to include the
  Merl header file:
         -include_lib("syntax_tools/include/merl.hrl").

  Then, you can use the ?Q(Text) macros in your code to create
  ASTs or match on existing ASTs. For example:
         Tuple = ?Q("{foo, 42}"),
         ?Q("{foo, _@Number}") = Tuple,
         Call = ?Q("foo:bar(_@Number)")

  Calling merl:print(Call) will then print the following code:
         foo:bar(42)

  The ?Q macros turn the quoted code fragments into ASTs, and
```

I'll take a look at fixing that and try not to break any of the erl_docgen renderings.

erszcz commented:
Thanks for the review, @garazdawi!

I think it makes the most sense to start with the spec/type changes. I'm sending a separate PR shortly, to make it easy to review. Then, I'll rebase this as appropriate.

erszcz commented on lib/edoc/doc/overview.edoc:
```diff
@@ -1093,6 +1101,37 @@ Details:
+  files => [],mode => chunks,run => app}
+$ ls _build/default/lib/recon/doc/chunks/
+recon.chunk        recon_lib.chunk    recon_rec.chunk
+recon_alloc.chunk  recon_map.chunk    recon_trace.chunk
```
I added a Rebar3 example and extended both steps to view docs in `erl`.

erszcz commented on lib/edoc/doc/overview.edoc:
```diff
@@ -1093,6 +1101,37 @@ Details:
+code documentation.
+
+There are two ways to generate the doc chunks: either by using `bin/edoc.escript'
+or by using the <a href="">Rebar3 plugin</a>. Ultimately, they're both just
```
Fixed, link to Rebar3 `edoc` is now in place, along with an example.

erszcz commented on lib/edoc/bin/edoc.escript:
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env escript
```
The script is now installed with the rest of OTP. I hope it's done properly - it seems to work fine. I also added flags to specify the output directory and include search paths.

erszcz commented:
With regard to `erl_syntax:tree/1,2`, I've added code which inserts argument names into the specs stored in the chunk. This means the spec in the chunk will not reflect the spec in the source file 1-to-1 anymore, but for the example from [`erl_syntax.erl`](https://github.com/erlang/otp/blob/28cd236d03d62faee5d3733a05e40453c768211f/lib/syntax_tools/src/erl_syntax.erl#L8173-L8176):

```erlang
-spec tree(atom()) -> tree().

tree(Type) ->
    tree(Type, []).

-spec tree(atom(), term()) -> tree().

tree(Type, Data) ->
    ...
```

We get properly named types in `erl`:

```
1> h(erl_syntax, tree).

  -spec tree(Type :: atom()) -> tree().

  Equivalent to tree(Type, []).

  -spec tree(Type :: atom(), Data :: term()) -> tree().

  For special purposes only. Creates an abstract syntax tree node with type tag Type and
  associated data Data.
```

It does not yet support 'bounded_fun' syntax tree nodes (`-spec f(A) -> A when A :: type()`) - for now, these are just passed as is.

erszcz commented on lib/edoc/src/edoc.erl:
```diff
@@ -60,23 +60,75 @@
 %% @equiv file(Name, [])
 %% @deprecated See {@link file/2} for details.
-
+-spec file(filename()) -> ok.
```
Please see https://github.com/erlang/otp/pull/2803#issuecomment-743219786

erszcz commented on lib/edoc/src/edoc.erl:
```diff
@@ -60,23 +60,75 @@
+-type ordset(T) :: ordsets:ordset(T).
+-type function_name() :: {atom(), integer()}.
+-type filename() :: file:filename().
+-type proplist() :: proplists:proplist().
```
I think they were just shorthands.

erszcz commented on lib/stdlib/src/shell_docs.erl:
```diff
@@ -31,6 +31,8 @@
 %% Convinience functions
 -export([get_doc/1, get_doc/3, get_type_doc/3, get_callback_doc/3]).
 
+-export_type([chunk_elements/0, chunk_element_attr/0]).
```
[It's used in a spec](https://github.com/erszcz/otp/blob/6e86552dffc1a5a2329958553374c1b6c83b0a42/lib/edoc/src/edoc_layout_chunks.erl#L383), since I wanted to highlight that the HTML tree produced by the layout chunk is later consumed by `shell_docs`. However, it's not critical, this type does not have to be exported - just let me know and I'll revert that.

erszcz commented:
> These should probably be grouped into one:
> ```
> See also data/1, is_tree/1, type/1
> ```

Done in the newest batch of changes.

Moreover, arg names are now properly added to `bounded_fun` kind of specs.

garazdawi commented:
Would you mind rebasing on top of latest master to get the #2914 changes removed from this PR?

erszcz commented:
@garazdawi Sure, I should be able to rebase this tomorrow. I also have some WIP on the documentation for the escript - I'll attach on top of the rebase.

garazdawi commented:
Great, let me know when you are ready for me to review the PR again.

erszcz commented:
I've rebased this on top of the current master. I think I've addressed all points of the first review - please let me know if there's something missing. The current status is as follows:

- `edoc.escript` is renamed to just `edoc` - I think it's simpler and better, if it's intended for public use. However, in the HTML table of contents it is listed as `edoc_cmd`, since otherwise it would clash with `edoc.erl` docs. I think it's fine this way, but let me know if you think otherwise.
- Specs stored in chunks don't reflect 1-to-1 the specs in source files - they're extended to contain arg names and type defs. I think that you, @garazdawi, intended to make the specs in source files and in the chunks to be 100% identical. Maybe EDoc should just warn "Hey, your spec looks like this: ..., but it'd look better like that: .... Would you mind fixing your source?" That would only require the user to copy-n-paste the rewritten example, but I'm afraid even that might be too much. This is borderline with EDoc being a linter. Opinions are welcome.
- @josevalim's good point about warnings for `@type` and `@spec`. This is slightly connected with the above point. Should this warning be part of this PR?
- If `@spec` and `@type` are considered deprecated, then what about deprecating `@deprecated` (pun intended ;) )? There's `xref` which supports a `-deprecated` attribute, so this case is analogous to `dialyzer` and `@spec`/`@type`. I think it's worth considering, but not necessarily as part of this PR. This would also solve the problem of EEP-48 defining the `deprecated` field as just a flat `binary()`, while current EDoc allowing it to contain markup.

TODOs:
- When testing this rebased code on Recon, I found out that `user_type` nodes are not handled in the spec rewriting code - I'll work on this next.

garazdawi commented:
Hello,

I did `lib/edoc/bin/edoc -chunks -app syntax_tools` and got a bunch of crashes looking like this:

```
edoc: error in layout 'edoc_layout_chunks': {'EXIT',
                                       {function_clause,
                                        [{edoc_layout_chunks,annotate_spec,
                                          [['Name'],undefined],
                                          [{file,"edoc_layout_chunks.erl"},
                                           {line,250}]},
```

Is that because they have both a `-spec` and `@spec`?

garazdawi commented:
> Specs stored in chunks don't reflect 1-to-1 the specs in source files - they're extended to contain arg names apart from types. I think that you, @garazdawi, intended to make the specs in source files and in the chunks to be 100% identical.

I think it is fine that they do not match 100%. This is an oddity in the way that edoc behaves, but I don't think it is worth forcing users to rewrite their specs.

garazdawi commented:
> warnings for @type and @spec. This is slightly connected with the above point. Should this warning be part of this PR?

I think it is better to do this in another PR.

garazdawi reviewed (commented):
I did a quick look to see if it would be easy to add support in the make system to use edoc instead of erl_docgen to build the doc chunks for Erlang/OTP and it turns out that it is not. So I think we will leave that for later.

My only remaining question from the original review is whether we should document the `edoc_doclet_chunks` and `edoc_layout_chunks` modules? The way it is implemented now it is only possible to generate chunks via the script without relying on undocumented API's.

erszcz commented on lib/edoc/src/edoc_layout_chunks.erl:
```diff
@@ -0,0 +1,512 @@
+	    AttrsNoDocgen = maps:without(['docgen-rel', 'docgen-href'], Attrs),
+	    NewAttrs = AttrsNoDocgen#{rel => expand_docgen_rel(ShortRel),
+				      href => URI},
+	    {Tag, [{'edoc-orig-tag', Tag}] ++ maps:to_list(NewAttrs), SubEls}
```
> What is the purpose of the 'edoc-orig-tag'?

EDoc accepts any HTML, so also `<a>` tags. Putting `edoc-orig-tag` in `<a>`s which are generated makes it easy to debug / test. I'm not sure if it could be of any use to the final consumers - I'll clean it up.

erszcz replied:
> The name attribute seems misplaced here...

Good catch, though this comes from EDoc core, not my changes. I'm removing it with the hope that nothing breaks 🤞

garazdawi replied:
> Good catch, though this comes from EDoc core, not my changes. I'm removing it with the hope that nothing breaks :crossed_fingers:

Then the HTML backend removes it somehow as well, as I don't think it ends up in the final HTML.

erszcz commented on lib/edoc/src/edoc_layout_chunks.erl:
```diff
@@ -0,0 +1,512 @@
+is_html_tag(Tag) ->
+    %% This is only a subset of existing HTML tags.
+    %% Compare with https://developer.mozilla.org/en-US/docs/Web/HTML/Element
+    Tags = [a,p,h1,h2,h3,i,br,em,pre,code,ul,ol,li,dl,dt,dd],
```
Hmm, from http://erlang.org/doc/apps/edoc/chapter.html#headings and your link it seems that heading ranks up to and including `h5` might be generated by EDoc, unless somebody explicitly uses `<h6>` tag. I think it would be cleaner to let them all pass through to maintain the document structure / nesting, but maybe just do not distinguish between the 4/5/6 levels in the shell renderer?

garazdawi replied:
yes, that makes sense. Can you add it to shell_docs, or should I?

erszcz replied:
I'll take care of it

garazdawi replied:
done in #2991

erszcz commented:
Status update:
> When testing this rebased code on Recon, I found out that user_type nodes (that is all non-builtin types) are not handled in the spec rewriting code - I'll work on this next.

Done.

---

> got a bunch of crashes ... because they have both a -spec and @spec?

Indeed, that turns out to be the reason. It seems that in such a case EDoc doesn't pass the -spec attr form down to the layout :| I might be able to find out why and fix it in EDoc core, but for now a warning will be printed that just `@spec` or redundant `-spec` and `@spec` are not supported when generating chunks. At least we have a clear answer to the "should we print a warning or not" question :)

garazdawi commented:
The only part missing from my point of view is the docs for the new layout and doclet. After that this should be ready to merge.

Btw, what deprecations/incompatibilities would you say this PR introduces? The `@spec` and `@type` annotations, anything else?

erszcz commented:
I've added the missing documentation.

Deprecations / incompatibilities:
- Definitely `@spec` and `@type`. Moreover, while the compiler doesn't care about `-spec` attribute order, EDoc does - it doesn't handle consecutive `-spec` attributes (since they're handled as though they were `@spec` tags, and these have to be interspersed with function definitions). This pops up as errors on some OTP files (e.g. `stdlib/src/otp_internal.erl`).
- When generating chunks both a `@spec` and `-spec` are found for the same function, EDoc will print a warning - this is also found in some OTP code.

Not necessarily deprecations/incompatibilities, but the following seem nice to haves in the future:
- In case of some multi-clause funs/specs, EDoc confuses the parameter names - there's an example logged by `eep48_SUITE` when running tests.
- EDoc allows `@deprecated` to contain markup, but it is completely flattened before putting into chunks to conform to EEP-48 (so, for example, no links in deprecation rationale).
- `@deprecated` and `@since` are not (and never were) allowed on type and callback definitions. There are intentionally skipped tests for these in `eep48_SUITE`.

I think that's all.

erszcz commented:
Well, it wasn't all :)
- It's rare, but some OTP libs (`compiler`, `syntax_tools`, `xmerl`) use HTML tables in their EDoc docs - table tags are stripped, but the contents are not, so the resulting documentation is probably a mess :|

garazdawi commented:
Should we add a header in the edoc user's guide about how to use the edoc API using doclet/layout to generate chunks? Also maybe add a link from the edoc_doclet/layout_chunks to that section in the user's guide?

Maybe also mention that it is [application/erlang+html](http://erlang.org/doc/apps/erl_docgen/doc_storage.html#erlang-documentation-format) that is generated?

garazdawi commented:
Does [this](http://erlang.org/doc/apps/edoc/chapter.html#type-specifications) documentation need updating?

```
Erlang specifications will be used unless there is also a function specification (a @spec tag followed by a type) with the same name.
```

erszcz commented:
I think all the important places are updated now. Please let me know if there's something missing.

garazdawi reviewed (changes requested):
All the tests seem to pass except one small issue. So I think it would be a good idea to cleanup the git history of the PR and then we can merge it.

garazdawi commented on lib/edoc/src/edoc.app.src:
```diff
@@ -1,26 +1,31 @@
   {env, []},
-  {runtime_dependencies, ["xmerl-1.3.7","syntax_tools-1.6.14","stdlib-2.5",
-			  "kernel-3.0","erts-6.0"]}]}.
+  {runtime_dependencies, ["xmerl-1.3.7", "syntax_tools-1.6.14", "stdlib-2.5",
```
```suggestion
  {runtime_dependencies, ["xmerl-1.3.7", "syntax_tools-1.6.14", "stdlib-@OTP-17120@",
```

Just realized that since we use functionality from `shell_docs` we need to update this.

richcarl commented:
I've not had enough bandwidth to look at this before, but I just want to state that I absolutely agree that documentation @spec and @type tags should cause a deprecation warning. Should have done that long ago.

richcarl commented on lib/edoc/src/edoc.app.src:
```diff
@@ -1,26 +1,31 @@
 {application, edoc,
- [{description, "EDoc"},
+ [
+  {description, "EDoc OTP library"},
```
I don't like this description. EDoc is not OTP and is not what you'd think of as a library.

erszcz replied:
Ahh, yes, this is one more thing that slipped in when I imported from the external repo...

richcarl replied:
I see. Just restore it in that case. :-) Looking at descriptions from other OTP apps, they tend to just have a single word or two.

erszcz replied:
I've modified the description. Please let me know, @richcarl, if it's appropriate now or if I should just leave it at "EDoc".

richcarl replied:
That's good, I like it!

erszcz commented:
> I just want to state that I absolutely agree that documentation @SPEC and @type tags should cause a deprecation warning.

Now that I've looked around a bit more, I think this warning should be coupled with actually giving the tags a lower priority than the attributes. Specifically, I mean addressing the comments and logic like this - https://github.com/erszcz/otp/blob/ea0046dfd49a2cae2eeea7d9a4e853d42740a76b/lib/edoc/src/edoc_specs.erl#L222-L231. I'd rather leave this for another PR, given we've reached a stable state of this one. @garazdawi @richcarl Is it ok?

richcarl commented:
> Now that I've looked around a bit more, I think this warning should be coupled with actually giving the tags a lower priority than the attributes. Specifically, I mean addressing the comments and logic like this - https://github.com/erszcz/otp/blob/ea0046dfd49a2cae2eeea7d9a4e853d42740a76b/lib/edoc/src/edoc_specs.erl#L222-L231. I'd rather leave this for another PR, given we've reached a stable state of this one. @garazdawi @richcarl Is it ok?

Sounds good to me.

garazdawi commented:
> Now that I've looked around a bit more, I think this warning should be coupled with actually giving the tags a lower priority than the attributes.

Yes, seems reasonable. If we end up breaking a lot of people's documentation builds we may have to patch in an option to switch it back to the old default, but we'll cross that bridge if we get to it.

> I'd rather leave this for another PR, given we've reached a stable state of this one. @garazdawi @richcarl Is it ok?

yes.

garazdawi commented:
Do you want to write the release note? This will be the 1.0 release of Edoc and it will be listed as a highlight (because this change is awesome) and as an incompatibility (because of the `-spec`/`@spec` inversion).

garazdawi commented:
github actions doc build has started to fail with the latest changes, do you know why?

erszcz commented:
https://github.com/erlang/otp/pull/2803/commits/ce77e68b296d126153f9938122662835e8d69b09 and https://github.com/erlang/otp/pull/2803/commits/c5581d27b0b4731480d04a444e926e7b28ab99b4 introduce the priority of attrs over tags. However, as this changes EDoc core, it's not limited to chunk generation, but all paths, so static HTML or `erl_docgen` spec extraction, too.

Specs were missing in OTP docs after these changes - 
https://github.com/erlang/otp/pull/2803/commits/93c34bb293dea2f759d54abb612eb9646639bc17 fixes that.

I'm now troubleshooting some types which used to be exported, but currently are not.

erszcz commented:
I’ve reverted the PR to the last commit passing all checks. I’ll carry on in a separate branch.

erszcz commented:
The latest commits invert the priorities of tags and attributes. `-spec` and `-type` now take precedence over `@spec` and `@type`. Warnings will be emitted for the tags. 

This priority inversion means that if a module has both a `@type` and a `-type` of the same name, but only the former carries a doc comment, then this comment will not appear anymore in either the HTML or chunk output. See `@type array` in `stdlib/src/array.erl` for an example. This is a backward incompatible change, but should not affect OTP docs.

In case of specs, the problem of consecutive `-spec` attributes is solved - EDoc can now accept specs in any order and not necessarily directly adjacent to a function they describe. This solves some errors in OTP files, e.g. `stdlib/src/otp_internal.erl`.

I finally think there's nothing more to add to this PR, unless some bugs pop up. @garazdawi @richcarl what do you think?

garazdawi commented:
> I finally think there's nothing more to add to this PR, unless some bugs pop up.

I agree.

garazdawi commented:
Merged! Thanks for all your work!

erszcz commented:
Woohoo! And thank you for the reviews!
